### PR TITLE
Support vLLM single and multi-host TPUs on GKE

### PR DESCRIPTION
**Pull request by richardsliu**
Fixes a few issues with vLLM on TPUs using GKE and RayServe:
- Installs ray dashboard and serve libraries in Docker image
- Calculates the number of nodes based on current placement group instead of the entire Ray cluster (which may include non-TPU nodes)
- Ensures TPU-specific environment variables are set correctly for each Ray worker

### CODE REVIEW

File: vllm/distributed/device_communicators/tpu_communicator.py [hunk 1]
```diff
@@ -24,9 +26,33 @@ def __init__(self, group: ProcessGroup):
...
+        num_nodes = len(nodes_in_pg)
+
         local_world_size = global_world_size // num_nodes
         local_rank = global_rank % local_world_size
+
+        print(f"global_rank: {global_rank}")
+        print(f"global_world_size: {global_world_size}")
+        print(f"num_nodes: {num_nodes}")
+        print(f"local_world_size: {local_world_size}")
+        print(f"local_rank: {local_rank}")
```
Comment (WoosukKwon) [inline vllm/distributed/device_communicators/tpu_communicator.py:50]: @richardsliu Thanks for submitting the PR! Can you please clean up these print statements?

Comment (richardsliu) [reply vllm/distributed/device_communicators/tpu_communicator.py]: @WoosukKwon Thanks, removed.

File: requirements-tpu.txt [hunk 2]
```diff
@@ -4,4 +4,4 @@
 # Dependencies for TPU
 # Currently, the TPU backend uses a nightly version of PyTorch XLA.
 # You can install the dependencies in Dockerfile.tpu.
-ray
+ray[default,serve]
```
Comment (WoosukKwon) [inline requirements-tpu.txt:7]: Please remove Ray Serve from the dependency. vLLM does not need to be used with Ray Serve, and we'd like to minimize the dependencies.

File: vllm/distributed/device_communicators/tpu_communicator.py [hunk 3]
```diff
@@ -24,9 +27,37 @@ def __init__(self, group: ProcessGroup):
...
+        # is the Ray placement group if it is deployed with Ray. Default
+        # to the number of TPU nodes in the Ray cluster. The number of TPU
+        # nodes is computed by the total number of TPUs divided by the
+        # number of TPU accelerators per node, to account for clusters
+        # with both CPUs and TPUs.
+        cluster_resources = ray.cluster_resources()
+        total_tpus = int(cluster_resources["TPU"])
+        tpus_per_node = (
+            TPUAcceleratorManager.get_current_node_num_accelerators())
+        num_nodes = total_tpus // tpus_per_node
```
Comment (WoosukKwon) [inline vllm/distributed/device_communicators/tpu_communicator.py:41]: nit: Can we add `assert total_tpus % tpus_per_node == 0`?

File: vllm/distributed/device_communicators/tpu_communicator.py [hunk 4]
```diff
@@ -24,9 +27,37 @@ def __init__(self, group: ProcessGroup):
...
+                    for _, node in pg["bundles_to_node_id"].items():
+                        nodes_in_pg.add(node)
+            num_nodes = len(nodes_in_pg)
+
         local_world_size = global_world_size // num_nodes
         local_rank = global_rank % local_world_size
+
+        # Ensure environment variables are set for multihost deployments.
+        os.environ["CLOUD_TPU_TASK_ID"] = str(global_rank)
+        os.environ["TPU_VISIBLE_CHIPS"] = str(local_rank)
```
Comment (WoosukKwon) [inline vllm/distributed/device_communicators/tpu_communicator.py:59]: Why is this needed?

File: vllm/distributed/device_communicators/tpu_communicator.py (hunk 3 above)
Comment (WoosukKwon) [reply vllm/distributed/device_communicators/tpu_communicator.py:41]: Can we move this to `ray_utils.py`?

File: vllm/executor/ray_tpu_executor.py [hunk 5]
```diff
@@ -70,6 +70,19 @@ def _init_workers_ray(self, placement_group: "PlacementGroup",
...
+            if "TPU_CHIPS_PER_HOST_BOUNDS" in os.environ:
+                override_env.update({
+                    "TPU_CHIPS_PER_HOST_BOUNDS":
+                    os.environ["TPU_CHIPS_PER_HOST_BOUNDS"]
+                })
+            if "TPU_HOST_BOUNDS" in os.environ:
+                override_env.update(
+                    {"TPU_HOST_BOUNDS": os.environ["TPU_HOST_BOUNDS"]})
+
             worker = ray.remote(
```
Comment (WoosukKwon) [inline vllm/executor/ray_tpu_executor.py:86]: QQ: can we use the `runtime_env` arg in `ray.remote` instead?

File: vllm/executor/ray_tpu_executor.py
Comment (WoosukKwon) [inline vllm/executor/ray_tpu_executor.py:1]: When TP=1 (e.g., v5e-1), vLLM does not use Ray at all and use `tpu_executor.py` (instead of `ray_tpu_executor`). Can this case also be supported in GKE?

**Review by WoosukKwon — COMMENTED** [review]
@richardsliu Thanks for the PR and sorry again for the delays in my review.

The PR looks good to me overall, but can be improved in terms of code organization. Please check out my comments.

File: requirements-tpu.txt (hunk 2 above)
Comment (richardsliu) [reply requirements-tpu.txt:7]: Can we still install ray[default]? The reason for this is that the GCS endpoint needs to run through the Ray dashboard which does not get installed if you just do `pip install ray`. The endpoint is needed in order for other Ray workers to join the cluster for multihost inference.

Comment (WoosukKwon) [reply requirements-tpu.txt:7]: Oh, how is `ray[default]` different from just `ray`?

Comment (richardsliu) [reply requirements-tpu.txt:7]: According to https://docs.ray.io/en/latest/ray-overview/installation.html, `ray[default]` includes the Ray dashboard while `ray` is just the Ray core libraries.

The reason for including the dashboard (in addition to debuggability) is that the GCS and Ray Job endpoints are exposed through the dashboard. So without it, the other Ray nodes are not able to join the cluster. For example the Kubernetes operator initializes Ray worker nodes by having them ping the GCS endpoint.

File: vllm/distributed/device_communicators/tpu_communicator.py (hunk 4 above)
Comment (richardsliu) [reply vllm/distributed/device_communicators/tpu_communicator.py:59]: This is needed for libtpu and TPU driver to know which TPU chip is actually visible. On GKE these need to be set, otherwise the TPU driver will fail to initialize because the number of devices would be different from the number of visible worker addresses.

**Review by WoosukKwon — APPROVED** [review]
@richardsliu LGTM! FYI, I made some minor modifications to the PR, mostly regarding style (I don't have preference, but just wanted to minimize the changes not related to this PR).
